File: Vista/NuevaOrdenDialog.py
from PyQt6.QtWidgets import (QDialog, QVBoxLayout, QHBoxLayout, QLabel, QLineEdit, QPushButton, 
                             QComboBox, QTableWidget, QTableWidgetItem, QHeaderView, QMessageBox, 
                             QWidget, QScrollArea, QInputDialog, QFrame)
from PyQt6.QtCore import Qt
from PyQt6.QtGui import QFont
from Vista.Estilos import Estilos
from Modelo.OrdenCompra import OrdenCompra, TipoPago, Estado as EstadoOrden
from datetime import datetime

class NuevaOrdenDialog(QDialog):
    def __init__(self, parent, ctrl):
        super().__init__(parent)
        self.ctrl = ctrl
        self.items = []
        self.setWindowTitle("Nueva Orden de Compra")
        self.setModal(True)
        self.construir()

    def construir(self):
        self.resize(740, 640)
        layout = QVBoxLayout(self)
        layout.setContentsMargins(0, 0, 0, 0)
        layout.setSpacing(0)

        panel = QWidget()
        panel.setStyleSheet(f"background-color: {Estilos.BG_BLANCO.name()};")
        p_layout = QVBoxLayout(panel)
        p_layout.setContentsMargins(0, 0, 0, 0)
        p_layout.setSpacing(0)

        header = QWidget()
        header.setStyleSheet(f"background-color: {Estilos.BG_ZINC_100.name()}; border-bottom: 1px solid {Estilos.BORDE.name()};")
        h_layout = QVBoxLayout(header)
        h_layout.setContentsMargins(24, 20, 24, 20)
        tit = QLabel("NUEVA ORDEN DE COMPRA")
        tit.setFont(Estilos.FUENTE_TITULO)
        tit.setStyleSheet("border: none;")
        h_layout.addWidget(tit)
        p_layout.addWidget(header)

        form = QWidget()
        form.setStyleSheet("border: none;")
        form_layout = QHBoxLayout(form)
        form_layout.setContentsMargins(24, 20, 24, 12)
        form_layout.setSpacing(24)

        izq = QWidget()
        izq_layout = QVBoxLayout(izq)
        izq_layout.setContentsMargins(0, 0, 0, 0)
        
        izq_layout.addWidget(self.lbl("1. Proveedor Responsable"))
        self.cmbProveedor = QComboBox()
        self.proveedores_list = self.ctrl.getProveedores()
        self.cmbProveedor.addItems([p.getNombre() for p in self.proveedores_list])
        self.cmbProveedor.setFont(Estilos.FUENTE_BOLD)
        self.cmbProveedor.setFixedHeight(42)
        izq_layout.addWidget(self.cmbProveedor)
        
        izq_layout.addSpacing(14)
        izq_layout.addWidget(self.lbl("2. Terminos de Pago"))
        self.cmbTipoPago = QComboBox()
        self.cmbTipoPago.addItems(["CONTADO", "CREDITO"])
        self.cmbTipoPago.setFont(Estilos.FUENTE_BOLD)
        self.cmbTipoPago.setFixedHeight(42)
        izq_layout.addWidget(self.cmbTipoPago)
        
        form_layout.addWidget(izq)

        der = QWidget()
        der_layout = QVBoxLayout(der)
        der_layout.setContentsMargins(0, 0, 0, 0)
        
        der_layout.addWidget(self.lbl("3. Buscar y Agregar Productos"))
        self.txtBusquedaProd = QLineEdit()
        self.txtBusquedaProd.setFont(Estilos.FUENTE_BOLD)
        self.txtBusquedaProd.setStyleSheet(f"""
            QLineEdit {{
                background-color: {Estilos.BG_ZINC_100.name()};
                border: 2px solid {Estilos.BORDE.name()};
                padding: 10px 12px;
            }}
        """)
        self.txtBusquedaProd.setFixedHeight(42)
        der_layout.addWidget(self.txtBusquedaProd)
        
        der_layout.addSpacing(8)
        btnAgregar = Estilos.botonExito("Agregar")
        btnAgregar.setFixedSize(120, 36)
        btnAgregar.clicked.connect(self.buscarYAgregar)
        der_layout.addWidget(btnAgregar)
        
        form_layout.addWidget(der)
        p_layout.addWidget(form)

        self.tablaItems = QTableWidget(0, 6)
        self.tablaItems.setHorizontalHeaderLabels(["Producto", "Unidad", "Cant.", "Costo Unit.", "Subtotal", "Quitar"])
        self.tablaItems.setFont(Estilos.FUENTE_NORMAL)
        self.tablaItems.horizontalHeader().setFont(Estilos.FUENTE_XS)
        self.tablaItems.horizontalHeader().setStyleSheet(f"background-color: {Estilos.BG_ZINC_100.name()};")
        self.tablaItems.setStyleSheet(f"background-color: {Estilos.BG_BLANCO.name()}; border: none;")
        self.tablaItems.verticalHeader().setDefaultSectionSize(40)
        
        self.tablaItems.horizontalHeader().setSectionResizeMode(0, QHeaderView.ResizeMode.Stretch)
        self.tablaItems.setColumnWidth(1, 120)
        self.tablaItems.setColumnWidth(2, 70)
        self.tablaItems.setColumnWidth(3, 110)
        self.tablaItems.setColumnWidth(4, 110)
        self.tablaItems.setColumnWidth(5, 90)
        
        self.tablaItems.itemDoubleClicked.connect(self.onTablaDoubleClick)
        
        p_layout.addWidget(self.tablaItems)

        footer = QWidget()
        footer.setStyleSheet(f"background-color: {Estilos.BG_ZINC_100.name()}; border-top: 1px solid {Estilos.BORDE.name()};")
        footer_layout = QHBoxLayout(footer)
        footer_layout.setContentsMargins(24, 16, 24, 16)
        
        self.lblTotal = QLabel("Inversion Total:  $0.00")
        self.lblTotal.setFont(QFont("SansSerif", 26, QFont.Weight.Bold))
        self.lblTotal.setStyleSheet(f"color: {Estilos.TEXTO_PRINCIPAL.name()}; border: none;")
        footer_layout.addWidget(self.lblTotal)
        
        footer_layout.addStretch()
        
        btnCancelar = Estilos.botonSecundario("Cancelar")
        btnCancelar.setFixedSize(110, 44)
        btnCancelar.clicked.connect(self.reject)
        footer_layout.addWidget(btnCancelar)
        
        btnConfirmar = Estilos.botonPrimario("Confirmar Orden")
        btnConfirmar.setFixedSize(180, 44)
        btnConfirmar.clicked.connect(self.confirmarOrden)
        footer_layout.addWidget(btnConfirmar)
        
        p_layout.addWidget(footer)
        layout.addWidget(panel)

    def onTablaDoubleClick(self, item):
        col = item.column()
        row = item.row()
        if col == 2:
            actual = self.tablaItems.item(row, 2).text()
            entrada, ok = QInputDialog.getText(self, "Nueva cantidad", "Nueva cantidad:", text=actual)
            if ok and entrada:
                try:
                    nueva = float(entrada.strip())
                    if nueva <= 0: return
                    item_orden = self.items[row]
                    if item_orden.getProducto().esPorPieza():
                        nueva = round(nueva)
                    item_orden.setCantidadSolicitada(nueva)
                    self.actualizarTablaItems()
                except ValueError:
                    pass

    def quitar_item_accion(self, row):
        if 0 <= row < len(self.items):
            self.items.pop(row)
            self.actualizarTablaItems()

    def buscarYAgregar(self):
        q = self.txtBusquedaProd.text().strip().lower()
        if not q: return
        
        res = []
        for p in self.ctrl.getProductos():
            if q in p.getNombre().lower():
                res.append(p)
            if len(res) >= 5:
                break
                
        if not res:
            QMessageBox.information(self, "Aviso", "No se encontraron productos.")
            return
            
        nombres = [p.getNombre() for p in res]
        sel_nombre, ok = QInputDialog.getItem(self, "Resultados", "Seleccione un producto:", nombres, 0, False)
        if not ok or not sel_nombre: return
        
        sel = next(p for p in res if p.getNombre() == sel_nombre)
        
        unidad = sel.getUnidad()
        mensajeCant = f"Cantidad a pedir (Piezas):" if sel.esPorPieza() else f"Cantidad a pedir ({unidad}):"
        entradaCant, ok2 = QInputDialog.getText(self, "Cantidad", mensajeCant, text="1" if sel.esPorPieza() else "1.0")
        if not ok2 or not entradaCant: return
        
        try:
            cantidad = float(entradaCant.strip())
            if cantidad <= 0: return
            if sel.esPorPieza():
                cantidad = round(cantidad)
        except ValueError:
            QMessageBox.critical(self, "Error", "Cantidad invalida")
            return
            
        for item in self.items:
            if item.getProducto().getId() == sel.getId():
                item.setCantidadSolicitada(item.getCantidadSolicitada() + cantidad)
                self.actualizarTablaItems()
                self.txtBusquedaProd.setText("")
                return
                
        # Inner class ItemOrden import requires nested usage or fully qualified
        from Modelo.OrdenCompra import ItemOrden
        self.items.append(ItemOrden(sel, cantidad, sel.getPrecio() * 0.75))
        self.actualizarTablaItems()
        self.txtBusquedaProd.setText("")

    def actualizarTablaItems(self):
        self.tablaItems.setRowCount(0)
        total = 0.0
        for i, item in enumerate(self.items):
            self.tablaItems.insertRow(i)
            p = item.getProducto()
            cantStr = f"{item.getCantidadSolicitada():.0f}" if p.esPorPieza() else f"{item.getCantidadSolicitada():.2f}"
            
            item0 = QTableWidgetItem(p.getNombre().upper())
            item0.setFlags(item0.flags() & ~Qt.ItemFlag.ItemIsEditable)
            self.tablaItems.setItem(i, 0, item0)
            
            item1 = QTableWidgetItem(p.getUnidad())
            item1.setFlags(item1.flags() & ~Qt.ItemFlag.ItemIsEditable)
            self.tablaItems.setItem(i, 1, item1)
            
            item2 = QTableWidgetItem(cantStr)
            item2.setFlags(item2.flags() & ~Qt.ItemFlag.ItemIsEditable)
            self.tablaItems.setItem(i, 2, item2)
            
            item3 = QTableWidgetItem(f"${item.getPrecioCosto():.2f}")
            item3.setFlags(item3.flags() & ~Qt.ItemFlag.ItemIsEditable)
            self.tablaItems.setItem(i, 3, item3)
            
            item4 = QTableWidgetItem(f"${item.getSubtotal():.2f}")
            item4.setFlags(item4.flags() & ~Qt.ItemFlag.ItemIsEditable)
            self.tablaItems.setItem(i, 4, item4)
            
            btnQuitar = QPushButton("Quitar")
            btnQuitar.setStyleSheet(f"background-color: {Estilos.ROSE_CLARO.name()}; color: {Estilos.ROSE.name()}; border: none; font-size: 11px;")
            btnQuitar.clicked.connect(lambda checked, row=i: self.quitar_item_accion(row))
            self.tablaItems.setCellWidget(i, 5, btnQuitar)
            
            total += item.getSubtotal()
            
        self.lblTotal.setText(f"Inversion Total:  ${total:.2f}")

    def confirmarOrden(self):
        if not self.items:
            QMessageBox.warning(self, "Aviso", "Agregue al menos un producto.")
            return
            
        idx = self.cmbProveedor.currentIndex()
        if idx < 0: return
        prov = self.proveedores_list[idx]
        
        tipoPago = TipoPago.CREDITO if self.cmbTipoPago.currentText() == "CREDITO" else TipoPago.CONTADO
        total = sum(item.getSubtotal() for item in self.items)
        
        orden = OrdenCompra(0, prov, list(self.items), total, tipoPago, EstadoOrden.PENDIENTE, datetime.now())
        self.ctrl.crearOrden(orden)
        QMessageBox.information(self, "Exito", f"Orden {orden.getFolioCorto()} generada correctamente.")
        self.accept()

    def lbl(self, t):
        l = QLabel(t.upper())
        l.setFont(Estilos.FUENTE_XS)
        l.setStyleSheet(f"color: {Estilos.TEXTO_TENUE.name()}; border: none;")
        return l

File: Vista/NuevaOrdenDialog.cs
using Microsoft.Maui.Controls;
using Microsoft.Maui.Graphics;
using Microsoft.Maui.Layouts;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using punto_de_venta_C_.Modelo;
using punto_de_venta_C_.Controlador;

namespace punto_de_venta_C_.Vista
{
    public class NuevaOrdenDialog : ContentPage
    {
        private readonly ProveedorControlador ctrl;
        private readonly List<OrdenCompra.ItemOrden> items = new List<OrdenCompra.ItemOrden>();
        private Picker cmbProveedor;
        private Picker cmbTipoPago;
        private Entry txtBusquedaProd;
        private VerticalStackLayout listItems;
        private Label lblTotal;
        private List<Proveedor> proveedoresList;
        private TaskCompletionSource<bool> tcs = new TaskCompletionSource<bool>();

        public NuevaOrdenDialog(ProveedorControlador ctrl)
        {
            this.ctrl = ctrl;
            Title = "Nueva Orden de Compra";
            BackgroundColor = Estilos.BG_BLANCO;
            construir();
        }

        private void construir()
        {
            var absoluteLayout = new AbsoluteLayout
            {
                WidthRequest = 740,
                HeightRequest = 640,
                BackgroundColor = Estilos.BG_BLANCO
            };

            var header = new Grid
            {
                BackgroundColor = Estilos.BG_ZINC_100,
                HeightRequest = 70,
                Padding = new Thickness(24, 15)
            };
            var tit = new Label
            {
                Text = "NUEVA ORDEN DE COMPRA",
                FontSize = Estilos.FUENTE_TITULO_SIZE,
                FontAttributes = FontAttributes.Bold,
                VerticalOptions = LayoutOptions.Center
            };
            header.Add(tit);
            AbsoluteLayout.SetLayoutBounds(header, new Rect(0, 0, 740, 70));
            AbsoluteLayout.SetLayoutFlags(header, AbsoluteLayoutFlags.None);
            absoluteLayout.Add(header);

            // Izquierda
            var lblProv = new Label { Text = "1. PROVEEDOR RESPONSABLE", FontSize = Estilos.FUENTE_XS_SIZE, TextColor = Estilos.TEXTO_TENUE };
            AbsoluteLayout.SetLayoutBounds(lblProv, new Rect(24, 90, 320, 14));
            AbsoluteLayout.SetLayoutFlags(lblProv, AbsoluteLayoutFlags.None);
            absoluteLayout.Add(lblProv);

            cmbProveedor = new Picker { FontAttributes = FontAttributes.Bold };
            proveedoresList = ctrl.getProveedores();
            foreach (var p in proveedoresList)
            {
                cmbProveedor.Items.Add(p.getNombre());
            }
            if (cmbProveedor.Items.Count > 0) cmbProveedor.SelectedIndex = 0;
            AbsoluteLayout.SetLayoutBounds(cmbProveedor, new Rect(24, 110, 320, 42));
            AbsoluteLayout.SetLayoutFlags(cmbProveedor, AbsoluteLayoutFlags.None);
            absoluteLayout.Add(cmbProveedor);

            var lblPago = new Label { Text = "2. TERMINOS DE PAGO", FontSize = Estilos.FUENTE_XS_SIZE, TextColor = Estilos.TEXTO_TENUE };
            AbsoluteLayout.SetLayoutBounds(lblPago, new Rect(24, 166, 320, 14));
            AbsoluteLayout.SetLayoutFlags(lblPago, AbsoluteLayoutFlags.None);
            absoluteLayout.Add(lblPago);

            cmbTipoPago = new Picker { FontAttributes = FontAttributes.Bold };
            cmbTipoPago.Items.Add("CONTADO");
            cmbTipoPago.Items.Add("CREDITO");
            cmbTipoPago.SelectedIndex = 0;
            AbsoluteLayout.SetLayoutBounds(cmbTipoPago, new Rect(24, 186, 320, 42));
            AbsoluteLayout.SetLayoutFlags(cmbTipoPago, AbsoluteLayoutFlags.None);
            absoluteLayout.Add(cmbTipoPago);

            // Derecha
            var lblBuscar = new Label { Text = "3. BUSCAR Y AGREGAR PRODUCTOS", FontSize = Estilos.FUENTE_XS_SIZE, TextColor = Estilos.TEXTO_TENUE };
            AbsoluteLayout.SetLayoutBounds(lblBuscar, new Rect(396, 90, 320, 14));
            AbsoluteLayout.SetLayoutFlags(lblBuscar, AbsoluteLayoutFlags.None);
            absoluteLayout.Add(lblBuscar);

            txtBusquedaProd = new Entry { FontSize = Estilos.FUENTE_BOLD_SIZE, BackgroundColor = Estilos.BG_ZINC_100 };
            txtBusquedaProd.Completed += (s, e) => buscarYAgregar();
            AbsoluteLayout.SetLayoutBounds(txtBusquedaProd, new Rect(396, 110, 320, 42));
            AbsoluteLayout.SetLayoutFlags(txtBusquedaProd, AbsoluteLayoutFlags.None);
            absoluteLayout.Add(txtBusquedaProd);

            var btnAgregar = Estilos.botonExito("Agregar");
            btnAgregar.Clicked += (s, e) => buscarYAgregar();
            AbsoluteLayout.SetLayoutBounds(btnAgregar, new Rect(396, 162, 120, 36));
            AbsoluteLayout.SetLayoutFlags(btnAgregar, AbsoluteLayoutFlags.None);
            absoluteLayout.Add(btnAgregar);

            // Table of selected products
            listItems = new VerticalStackLayout();
            var scroll = new ScrollView
            {
                Content = listItems,
                BackgroundColor = Estilos.BG_BLANCO
            };
            AbsoluteLayout.SetLayoutBounds(scroll, new Rect(24, 250, 692, 280));
            AbsoluteLayout.SetLayoutFlags(scroll, AbsoluteLayoutFlags.None);
            absoluteLayout.Add(scroll);

            // Footer
            var footer = new Grid
            {
                BackgroundColor = Estilos.BG_ZINC_100,
                HeightRequest = 80,
                Padding = new Thickness(24, 15),
                ColumnDefinitions = new ColumnDefinitionCollection
                {
                    new ColumnDefinition { Width = GridLength.Star },
                    new ColumnDefinition { Width = 300 }
                }
            };

            lblTotal = new Label
            {
                Text = "Inversion Total:  $0.00",
                FontSize = 24,
                FontAttributes = FontAttributes.Bold,
                TextColor = Estilos.TEXTO_PRINCIPAL,
                VerticalOptions = LayoutOptions.Center
            };
            footer.Add(lblTotal, 0, 0);

            var btns = new HorizontalStackLayout
            {
                Spacing = 10,
                HorizontalOptions = LayoutOptions.End,
                VerticalOptions = LayoutOptions.Center
            };

            var btnCancelar = Estilos.botonSecundario("Cancelar");
            btnCancelar.HeightRequest = 44;
            btnCancelar.WidthRequest = 110;
            btnCancelar.Clicked += async (s, e) =>
            {
                tcs.SetResult(false);
                await Navigation.PopModalAsync();
            };
            btns.Add(btnCancelar);

            var btnConfirmar = Estilos.botonPrimario("Confirmar Orden");
            btnConfirmar.HeightRequest = 44;
            btnConfirmar.WidthRequest = 180;
            btnConfirmar.Clicked += (s, e) => confirmarOrden();
            btns.Add(btnConfirmar);

            footer.Add(btns, 1, 0);

            AbsoluteLayout.SetLayoutBounds(footer, new Rect(0, 560, 740, 80));
            AbsoluteLayout.SetLayoutFlags(footer, AbsoluteLayoutFlags.None);
            absoluteLayout.Add(footer);

            Content = new Grid
            {
                Children = { absoluteLayout },
                HorizontalOptions = LayoutOptions.Center,
                VerticalOptions = LayoutOptions.Center
            };

            actualizarTablaItems();
        }

        private async void buscarYAgregar()
        {
            string q = txtBusquedaProd.Text?.Trim() ?? "";
            if (string.IsNullOrEmpty(q))
                return;

            List<Producto> res = new List<Producto>();
            foreach (var p in ctrl.getProductos())
            {
                if (p.getNombre().Contains(q, StringComparison.OrdinalIgnoreCase))
                    res.Add(p);
                if (res.Count >= 5)
                    break;
            }

            if (res.Count == 0)
            {
                await DisplayAlert("Aviso", "No se encontraron productos.", "OK");
                return;
            }

            string[] names = res.Select(p => p.getNombre()).ToArray();
            string choice = await DisplayActionSheet("Seleccione un producto:", "Cancelar", null, names);
            if (string.IsNullOrEmpty(choice) || choice == "Cancelar")
                return;

            Producto sel = res.First(p => p.getNombre() == choice);
            string unidad = sel.getUnidad();
            string mensajeCant = sel.esPorPieza() ? "Cantidad a pedir (Piezas):" : "Cantidad a pedir (" + unidad + "):";
            string entradaCant = await DisplayPromptAsync("Agregar Producto", mensajeCant, "Aceptar", "Cancelar", sel.esPorPieza() ? "1" : "1.0");

            if (string.IsNullOrEmpty(entradaCant))
                return;

            double cantidad;
            if (!double.TryParse(entradaCant.Trim(), out cantidad) || cantidad <= 0)
            {
                await DisplayAlert("Error", "Cantidad invalida", "OK");
                return;
            }
            if (sel.esPorPieza())
                cantidad = Math.Round(cantidad);

            foreach (var item in items)
            {
                if (item.getProducto().getId().Equals(sel.getId()))
                {
                    item.setCantidadSolicitada(item.getCantidadSolicitada() + cantidad);
                    actualizarTablaItems();
                    txtBusquedaProd.Text = "";
                    return;
                }
            }

            items.Add(new OrdenCompra.ItemOrden(sel, cantidad, sel.getPrecio() * 0.75));
            actualizarTablaItems();
            txtBusquedaProd.Text = "";
        }

        private void actualizarTablaItems()
        {
            listItems.Children.Clear();

            var header = new Grid
            {
                BackgroundColor = Estilos.BG_ZINC_100,
                HeightRequest = 30,
                ColumnDefinitions = new ColumnDefinitionCollection
                {
                    new ColumnDefinition { Width = GridLength.Star },
                    new ColumnDefinition { Width = 100 },
                    new ColumnDefinition { Width = 80 },
                    new ColumnDefinition { Width = 110 },
                    new ColumnDefinition { Width = 110 },
                    new ColumnDefinition { Width = 90 }
                }
            };
            header.Add(new Label { Text = "Producto", FontAttributes = FontAttributes.Bold, VerticalOptions = LayoutOptions.Center, Margin = new Thickness(5, 0) }, 0);
            header.Add(new Label { Text = "Unidad", FontAttributes = FontAttributes.Bold, VerticalOptions = LayoutOptions.Center }, 1);
            header.Add(new Label { Text = "Cant.", FontAttributes = FontAttributes.Bold, VerticalOptions = LayoutOptions.Center }, 2);
            header.Add(new Label { Text = "Costo Unit.", FontAttributes = FontAttributes.Bold, VerticalOptions = LayoutOptions.Center }, 3);
            header.Add(new Label { Text = "Subtotal", FontAttributes = FontAttributes.Bold, VerticalOptions = LayoutOptions.Center }, 4);
            header.Add(new Label { Text = "Quitar", FontAttributes = FontAttributes.Bold, VerticalOptions = LayoutOptions.Center }, 5);
            listItems.Children.Add(header);

            double total = 0;
            for (int i = 0; i < items.Count; i++)
            {
                int index = i;
                var item = items[i];
                var p = item.getProducto();

                var rowGrid = new Grid
                {
                    HeightRequest = 40,
                    ColumnDefinitions = new ColumnDefinitionCollection
                    {
                        new ColumnDefinition { Width = GridLength.Star },
                        new ColumnDefinition { Width = 100 },
                        new ColumnDefinition { Width = 80 },
                        new ColumnDefinition { Width = 110 },
                        new ColumnDefinition { Width = 110 },
                        new ColumnDefinition { Width = 90 }
                    }
                };

                rowGrid.Add(new Label { Text = p.getNombre().ToUpper(), VerticalOptions = LayoutOptions.Center, Margin = new Thickness(5, 0) }, 0);
                rowGrid.Add(new Label { Text = p.getUnidad(), VerticalOptions = LayoutOptions.Center }, 1);

                // Entry for editable quantity
                var entryCant = new Entry
                {
                    Text = p.esPorPieza() ? item.getCantidadSolicitada().ToString("0") : item.getCantidadSolicitada().ToString("0.00"),
                    Keyboard = Keyboard.Numeric,
                    VerticalOptions = LayoutOptions.Center,
                    HeightRequest = 32
                };
                entryCant.Unfocused += (s, e) =>
                {
                    if (double.TryParse(entryCant.Text, out double nv) && nv > 0)
                    {
                        if (p.esPorPieza()) nv = Math.Round(nv);
                        item.setCantidadSolicitada(nv);
                        actualizarTablaItems();
                    }
                    else
                    {
                        entryCant.Text = p.esPorPieza() ? item.getCantidadSolicitada().ToString("0") : item.getCantidadSolicitada().ToString("0.00");
                    }
                };
                rowGrid.Add(entryCant, 2);

                rowGrid.Add(new Label { Text = string.Format("${0:0.00}", item.getPrecioCosto()), VerticalOptions = LayoutOptions.Center }, 3);
                rowGrid.Add(new Label { Text = string.Format("${0:0.00}", item.getSubtotal()), VerticalOptions = LayoutOptions.Center }, 4);

                var btnQuitar = Estilos.botonPeligro("Quitar");
                btnQuitar.FontSize = Estilos.FUENTE_XS_SIZE;
                btnQuitar.HeightRequest = 28;
                btnQuitar.Padding = new Thickness(0);
                btnQuitar.Clicked += (s, e) =>
                {
                    items.RemoveAt(index);
                    actualizarTablaItems();
                };
                rowGrid.Add(btnQuitar, 5);

                listItems.Children.Add(rowGrid);
                total += item.getSubtotal();
            }

            lblTotal.Text = string.Format("Inversion Total:  ${0:0.00}", total);
        }

        private async void confirmarOrden()
        {
            if (items.Count == 0)
            {
                await DisplayAlert("Aviso", "Agregue al menos un producto.", "OK");
                return;
            }

            if (cmbProveedor.SelectedIndex < 0)
            {
                await DisplayAlert("Aviso", "Seleccione un proveedor.", "OK");
                return;
            }

            Proveedor prov = proveedoresList[cmbProveedor.SelectedIndex];
            OrdenCompra.TipoPago tipoPago = cmbTipoPago.SelectedIndex == 1 ? OrdenCompra.TipoPago.CREDITO : OrdenCompra.TipoPago.CONTADO;

            double total = items.Sum(item => item.getSubtotal());
            OrdenCompra orden = new OrdenCompra(0, prov, new List<OrdenCompra.ItemOrden>(items), total, tipoPago,
                    OrdenCompra.Estado.PENDIENTE, DateTime.Now);

            ctrl.crearOrden(orden);
            await DisplayAlert("Éxito", "Orden " + orden.getFolioCorto() + " generada correctamente.", "OK");
            tcs.SetResult(true);
            await Navigation.PopModalAsync();
        }

        public async Task<bool> mostrar()
        {
            if (Application.Current?.MainPage != null)
            {
                await Application.Current.MainPage.Navigation.PushModalAsync(new NavigationPage(this));
            }
            return await tcs.Task;
        }
    }
}
